### PR TITLE
Cache options we compute during refactoring operations (option2)

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This approach differs from https://github.com/dotnet/roslyn/pull/73725 in that we don't use a CWT.  We stoer the data as an instnace member on StructuredAnalyzerConfigOptions

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Diagnostics/StructuredAnalyzerConfigOptions.Implementation.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        public override IEnumerable<string> Keys
+            => _options.Keys;
+
+        public override NamingStylePreferences GetNamingStylePreferences()
+            => _lazyNamingStylePreferences.Value;
+
+#if !CODE_STYLE
+        public override CodeGenerationOptions GetCodeGenerationOptions(LanguageServices languageServices, CodeGenerationOptions? fallbackOptions)
+        {
+            _codeGenerationOptions ??= ((IOptionsReader)this).GetCodeGenerationOptions(languageServices, fallbackOptions);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Diagnostics/StructuredAnalyzerConfigOptions.Implementation.cs:47]: here we defer to the existing extension method to go do the work, which we then cache on this instance.

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Diagnostics/StructuredAnalyzerConfigOptions.Implementation.cs:47]: an alternate approach to this is in: https://github.com/dotnet/roslyn/pull/73725

**Review by tmat — APPROVED** [review]

**Comment by tmat** [comment]
This approach is preferable to CWT - the NamingStylePreferences already use the same caching strategy.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Diagnostics/StructuredAnalyzerConfigOptions.Implementation.cs (hunk 1 above)
Comment (tmat) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Diagnostics/StructuredAnalyzerConfigOptions.Implementation.cs:47]: Need to check if it is correct to pass fallback options here...

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Diagnostics/StructuredAnalyzerConfigOptions.Implementation.cs]: How would i check this?  Doesn't 'fallbackOptions' mean 'the default options if nothing is specified'?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Diagnostics/StructuredAnalyzerConfigOptions.Implementation.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        public override CodeGenerationOptions GetCodeGenerationOptions(LanguageServices languageServices, CodeGenerationOptions? fallbackOptions)
+        {
+            ref var options =
+                ref languageServices.Language == LanguageNames.CSharp ? ref _csharpCodeGenerationOptions :
+                ref languageServices.Language == LanguageNames.VisualBasic ? ref _visualBasicCodeGenerationOptions : ref Unsafe.NullRef<CodeGenerationOptions?>();
+
+            Contract.ThrowIfNull(options);
+
+            options ??= ((IOptionsReader)this).GetCodeGenerationOptions(languageServices, fallbackOptions);
+            return options;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Diagnostics/StructuredAnalyzerConfigOptions.Implementation.cs:57]:
```suggestion
            if (languageServices.Language == LanguageNames.CSharp)
                return _csharpCodeGenerationOptions ??= ((IOptionsReader)this).GetCodeGenerationOptions(languageServices, fallbackOptions);

            if (languageServices.Language == LanguageNames.VisualBasic)
                return _visualBasicCodeGenerationOptions ??= ((IOptionsReader)this).GetCodeGenerationOptions(languageServices, fallbackOptions);

            throw ExceptionUtilities.Unreachable();
```

**Pull request closed without merging**